### ISSUE TITLE
Classnames of the exported TensorRT model doesn't match the pytorch model

### ISSUE THREAD

**Issue by alikaz3mi** [ultralytics/yolov5#6328]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Export

### Bug

Hello. I have a model with 4 classes. When I exported the model as TensorRT, model.names displayed 1000 classes names `class0` to `class999`. Could you help me with this issue?

### Minimal Reproducible Example

```python
python export.py --include engine --device 0 --half --weights best.pt
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by dariogonle** [comment]
Hi @alikaz3mi, I think youu have to add the --data option. For example:
```
python export.py --include engine --device 0 --half --weights best.pt --data data.yaml
```

**Comment by glenn-jocher** [comment]
@alikaz3mi yes you can add a data.yaml path to detect.py to read class names from that yaml. I think otherwise they will default to COCO128 class names. You may need to update your code to see this feature as it's recent.
https://github.com/ultralytics/yolov5/blob/fd55271c04e9be68ea5299f7fe2aafcf4dc1984d/detect.py#L220

**Comment by dsoltysiuk** [comment]
@glenn-jocher This does not help, classes are still named `class0`-`class998`
The intermediate ONNX model (the one exported from `pt` format and will be converted to `TensorRT`) though has correct class names, meaning `export.py` was able to correctly parse `data.yaml`

**Comment by glenn-jocher** [comment]
@dsoltysiuk i apologize for not understanding the issue correctly earlier. It seems that the class names are correct in the intermediate ONNX model but they are not exported correctly in the TensorRT model. This is a known issue and is being actively worked on. We appreciate your feedback and please let us know if you have any additional concerns or questions in the meantime. Thank you for using YOLOv5.

**Comment by dsoltysiuk** [comment]
@glenn-jocher Thank you, great to hear that! I actually have one question regarding the image shape that can be passed to the TensorRT model. I tried using `--dynamic` flag, which in theory should allow different input/output shapes, but still getting `input size torch.Size([1, 3, 738, 1280]) not equal to max model size (1, 3, 768, 1280).` (there are already some issues presented, like this https://github.com/ultralytics/yolov5/issues/11033). 

I also don't really understand this error message; if the max model size is 1280x768 and my image resolution is 1280x738, why is it not accepted if it fits into the specified max resolution?

**Comment by glenn-jocher** [comment]
@dsoltysiuk, glad to hear the first issue is resolved! Regarding the issue with the `--dynamic` flag, let me try to clarify the behavior: `--dynamic` enables dynamic input shapes for TensorRT, not dynamic inference (i.e., variable number of detections per image). Thus you still need to define the maximum resolution that TensorRT should expect using `--max_batch_size` (default `1`) and `--max_shape`. 

Currently, the `--dynamic` flag has limited functionality. If you want to use the dynamic input shapes feature, but still get error messages such as "input size not equal to max model size", you may try adding `--explicit_batch`, which may resolve the mismatch issue.

Regarding the specific error message you are seeing: even though your image resolution fits into the max resolution, TensorRT requires it to also fit the tensor shapes specified by the Nvidia TRT API function `BuilderIOPlugin::getTensorRTRequiredDims()`, which might have some padding (i.e., in multiples of 8,16, 32, etc.). So while your image resolution is within range, the number of channels or batch size may not match the specified shapes. So it's hard to say exactly why your input is invalid without knowing more of the model architecture and specification.

Hope this helps, and feel free to ask if you have more questions. Let us know if this solves the issue you are facing!

**Comment by dsoltysiuk** [comment]
Thank you for the answer, @glenn-jocher 🙃
Explicit batch is enabled by default in YOLOv5 TensorRT export: https://github.com/ultralytics/yolov5/blob/0004c745daeebb0eb6455b90c3e8abb308e61320/export.py#L347

So I still cannot feed my model with different picture sizes 😔

**Comment by glenn-jocher** [comment]
@dsoltysiuk, thank you for clarifying the export flag. It seems that the `--dynamic` flag is still not fully functioning correctly for input shapes different than the default YOLOv5 size. 

We understand the importance of passing different image sizes to the model and know that this is a high priority for our users. Our team is working hard to make this possible with TensorRT as well. In the meantime, there may be workarounds to input images with different sizes, such as padding the smaller ones to match the largest. This issue will be updated as soon as we have information on progress towards a solution.

**Comment by dsoltysiuk** [comment]
Fantastic. Sorry for requesting too much from you guys...
Thank you for your quick responses and dedication 🙂
Have a productive day ☺️

**Comment by glenn-jocher** [comment]
@dsoltysiuk you're very welcome! We appreciate your understanding and support. Our team is always here to help and we're dedicated to providing the best solutions we can. If you have any more questions or concerns, don't hesitate to reach out. Have a productive day! 😊

**Comment by MOXHAN** [comment]
Not sure if this is needed or obvious information, but I just used this to resize any image

`transform = transforms.Compose([ transforms.CenterCrop(640) ])`
`img_transform = transform(img)`

**Comment by MOXHAN** [comment]
> @dsoltysiuk i apologize for not understanding the issue correctly earlier. It seems that the class names are correct in the intermediate ONNX model but they are not exported correctly in the TensorRT model. This is a known issue and is being actively worked on. We appreciate your feedback and please let us know if you have any additional concerns or questions in the meantime. Thank you for using YOLOv5.

Is there a work around for this? Im directly loading the yolov5s.engine from torch.hub and unfortunately only get class names like "class06"
Kind regards :)

**Comment by glenn-jocher** [comment]
@MOXHAN the issue you're experiencing with the class names in the TensorRT model is a known problem and our team is actively working on it. Currently, there is no official workaround for this issue. We apologize for the inconvenience caused and appreciate your patience. If you have any further questions or concerns, please feel free to let us know. Thank you for your understanding and support.

**Comment by MOXHAN** [comment]
Alright, thanks for your quick answer :)

I have found a workaround in the meantime. Its possible to simply load the labels into a pandas DataFrame and then overwrite the names of classes, which are accessible through model.names like this:
```
labels = pd.read_csv("labels.txt",header=None)

for i in range(len(model.names)):
    model.names[i] = labels[0][i]
```

However, I encountered another "problem": The normal Yolov5s model has the 80 COCO classes, but the TensorRT engine has 1000 (found through model.names, but returns "class0" and so on). Where can I find the real labels?

Kind regards and thanks for your help :)

**Comment by glenn-jocher** [comment]
Thanks for sharing your workaround, @MOXHAN! Using a pandas DataFrame to overwrite the class names is indeed a clever solution to the current issue.

Regarding the additional problem you encountered with the TensorRT engine having 1000 classes instead of 80, this discrepancy can be attributed to the fact that the TensorRT engine is initialized with default anchor boxes. These anchor boxes are used during inference to generate the bounding box predictions. However, the model.names list contains the original class names from the YOLOv5s model, which only has 80 classes.

To access the real labels for the 1000 classes in the TensorRT engine, you may need to refer to the specific dataset or use-case for which the engine was trained. Unfortunately, we don't have a readily available source for these labels.

If you have any more questions or need further assistance, please feel free to ask. We're here to help!

Kind regards and thanks for your contribution :)

**Comment by EvelinaAlexiutenko** [comment]
I found the solution, If you pass in the data argument file with your class labels In the same format as coco128.yaml you will get the right class labels in the result. As an example:
```
names:
  0: <your_class_label_0>
  1: <your_class_label_1>
```

**Comment by EvelinaAleksiutenko** [comment]
> @dsoltysiuk i apologize for not understanding the issue correctly earlier. It seems that the class names are correct in the intermediate ONNX model but they are not exported correctly in the TensorRT model. This is a known issue and is being actively worked on. We appreciate your feedback and please let us know if you have any additional concerns or questions in the meantime. Thank you for using YOLOv5.

I found the solution, If you pass in the data argument file with your class labels In the same format as coco128.yaml you will get the right class labels in the result. As an example:
```
names:
  0: <your_class_label_0>
  1: <your_class_label_1>
```

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution, @EvelinaAlexiutenko! It's great to hear that you have found a way to get the correct class labels in the result.

By passing in a data argument file with your class labels in the same format as coco128.yaml, you can ensure that the exported TensorRT model has the right class labels. This can be achieved by specifying the class labels in the `names` section of the data file.

Your example demonstrates how to define the class labels using the index-value mapping within the `names` section.

Thank you for contributing to the YOLOv5 community by sharing your solution. Should you have any further questions or require additional assistance, please don't hesitate to reach out.
